util/topsql/reporter/reporter.go

```diff
@@ -523,11 +519,53 @@ func (tsr *RemoteTopSQLReporter) getReportData(collected collectedData) ReportDa
 		records = append(records, others)
 	}
 
-	return ReportData{
-		collectedData:     records,
-		normalizedSQLMap:  collected.normalizedSQLMap,
-		normalizedPlanMap: collected.normalizedPlanMap,
+	return
+}
+
+// buildReportData convert record data in dataPoints slice and meta data in sync.Map to ReportData.
+//
+// Attention, caller should guarantee no more reader or writer access `sqlMap` and `planMap`, because buildReportData
+// will do heavy jobs in sync.Map.Range and it may block other readers and writers.
+func (tsr *RemoteTopSQLReporter) buildReportData(records []*dataPoints, sqlMap *sync.Map, planMap *sync.Map) ReportData {
+	res := ReportData{
+		CPUTimeRecords: make([]*tipb.CPUTimeRecord, 0, len(records)),
+		SQLMetas:       make([]*tipb.SQLMeta, 0, len(records)),
+		PlanMetas:      make([]*tipb.PlanMeta, 0, len(records)),
 	}
+
+	for _, record := range records {
+		res.CPUTimeRecords = append(res.CPUTimeRecords, &tipb.CPUTimeRecord{
+			RecordListTimestampSec: record.TimestampList,
+			RecordListCpuTimeMs:    record.CPUTimeMsList,
+			SqlDigest:              record.SQLDigest,
+			PlanDigest:             record.PlanDigest,
+		})
+	}
+
+	sqlMap.Range(func(key, value interface{}) bool {
+		meta := value.(SQLMeta)
+		res.SQLMetas = append(res.SQLMetas, &tipb.SQLMeta{
+			SqlDigest:     []byte(key.(string)),
```

Comment:
is it by design that they are stored as binary strings?

Reply:
Don't know. It has been like that for a long while.